File: templates/python/power_factor_locus.py
"""power_factor_locus: 不同负载下功率因数随时间变化."""
import sys, os; sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', '..', '_utils', 'python'))
import numpy as np
import matplotlib.pyplot as plt
from theme import apply_theme
from palette import cycle

def make_figure(title='Power factor over day'):
    apply_theme()
    h = np.linspace(0, 24, 200)
    pf_resi = 0.95 - 0.05*np.sin((h-12)*np.pi/12)
    pf_indu = 0.75 + 0.15*np.sin((h-12)*np.pi/12)
    pf_comm = 0.85 + 0.08*np.cos((h-14)*np.pi/12)
    fig, ax = plt.subplots()
    for i, (lab, pf) in enumerate([('residential', pf_resi), ('industrial', pf_indu),
                                    ('commercial', pf_comm)]):
        ax.plot(h, pf, color=cycle(i), label=lab)
    ax.axhline(0.9, color='red', linestyle='--', linewidth=0.7, label='target ≥ 0.9')
    ax.set_xlabel('hour'); ax.set_ylabel('power factor'); ax.set_title(title)
    ax.set_xticks(range(0, 25, 3)); ax.legend()
    ax.grid(True, linestyle=':', alpha=0.5)
    fig.tight_layout()
    return fig

if __name__ == '__main__':
    fig = make_figure(); fig.savefig(__file__.replace('.py', '.png'), dpi=150)

File: templates/matlab/power_factor_locus.m
function fig = power_factor_locus()
    addpath(fullfile(fileparts(mfilename('fullpath')),'..','..','_utils','matlab'));
    apply_theme();
    h = linspace(0, 24, 200);
    fig = figure;
    plot(h, 0.95 - 0.05*sin((h-12)*pi/12), 'Color', palette('cat',1)); hold on;
    plot(h, 0.75 + 0.15*sin((h-12)*pi/12), 'Color', palette('cat',2));
    plot(h, 0.85 + 0.08*cos((h-14)*pi/12), 'Color', palette('cat',3));
    yline(0.9, '--r', 'target \geq 0.9');
    xlabel('hour'); ylabel('power factor'); title('Power factor over day');
    legend({'residential','industrial','commercial'}); xticks(0:3:24); grid on;
end
